normalized-cache-incubating/src/commonMain/kotlin/com/apollographql/cache/normalized/internal/DefaultApolloStore.kt

```diff
@@ -118,11 +192,111 @@ internal class DefaultApolloStore(
         cacheKey = CacheKey.rootKey(),
         variables = variables,
         fieldKeyGenerator = fieldKeyGenerator,
+        returnPartialResponses = true,
     )
-    return ReadResult(
-        data = batchReaderData.toData(operation.adapter(), customScalarAdapters, variables),
-        cacheHeaders = batchReaderData.cacheHeaders,
-    )
+    val dataWithErrors: Map<String, Any?> = batchReaderData.toMap()
+    val errors = mutableListOf<Error>()
+
+    @Suppress("UNCHECKED_CAST")
+    val dataWithNullBubbling: Map<String, Any?>? = nullBubble(dataWithErrors, operation.rootField(), errors) as Map<String, Any?>?
+    val falseVariablesCustomScalarAdapter =
+      customScalarAdapters.newBuilder().falseVariables(variables.valueMap.filter { it.value == false }.keys).build()
+    val data = dataWithNullBubbling?.let { operation.adapter().fromJson(it.jsonReader(), falseVariablesCustomScalarAdapter) }
+    return ApolloResponse.Builder(operation, uuid4())
+        .data(data)
+        .errors(errors.takeIf { it.isNotEmpty() })
+        .cacheHeaders(batchReaderData.cacheHeaders)
+        .cacheInfo(
+            CacheInfo.Builder()
+                .fromCache(true)
+                .cacheHit(errors.isEmpty())
+                .stale(batchReaderData.cacheHeaders.headerValue(ApolloCacheHeaders.STALE) == "true")
+                .build()
+        )
+        .build()
+  }
+
+  /**
+   * If a position contains an Error, replace it by a null if the field's type is nullable, bubble up if not.
+   */
+  private fun nullBubble(dataWithErrors: Any?, field: CompiledField, errors: MutableList<Error>): Any? {
+    return when (dataWithErrors) {
+      is Map<*, *> -> {
+        @Suppress("UNCHECKED_CAST")
+        dataWithErrors as Map<String, Any?>
+        dataWithErrors.mapValues { (key, value) ->
+          val selection = field.fieldSelections()[key]
+              ?: // Scalar
+              return@mapValues value
```

Comment:
Not 100% sure but shouldn't there be a check for null in the scalar case as well?

Reply:
This is the _inside_ values of a scalar, which we shouldn't touch, so I think this is good (unless I'm missing something!). I added a test to be sure.

Reply:
Ah yes! Then I think you can lift the logic outside the `mapValues {}` loop? 

```kotlin
        val subSelections = field.fieldSelections()
        if (subSelections.isEmpty()) {
          // This is a scalar represented as a Map.
          return dataWithErrors
        }
        dataWithErrors.mapValues { (key, value) ->
          val selection = subSelections[key]
          if (selection == null) {
            // Should never happen
            return@mapValues value
          }
```

Reply:
Yes, much better 👍